Use span for MappedFileData and DataSegment

darinadler opened the pull request:
#### 6646f0585b449208fb9eabe1b43a3dd5ae153e6f
<pre>
Use span for MappedFileData and DataSegment
<a href="https://bugs.webkit.org/show_bug.cgi?id=274756">https://bugs.webkit.org/show_bug.cgi?id=274756</a>
<a href="https://rdar.apple.com/128807880">rdar://128807880</a>

Reviewed by Chris Dumez.

Changed JSC::CachePayload, WTF::MappedFileData, and
WebCore::DataSegment::Provider to use spans instead of pointer/size pairs.

* Source/JavaScriptCore/runtime/CachePayload.cpp:
(JSC::CachePayload::span const): Added.
(JSC::CachePayload::data const): Deleted.
(JSC::CachePayload::size const): Deleted.
* Source/JavaScriptCore/runtime/CachePayload.h: Updated for the above.

* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Decoder::offsetOf): Tweaked coding style a bit
(JSC::decodeCodeBlockImpl): Ditto.
(JSC::isCachedBytecodeStillValid): Removed use of CachePayload::size
in function that already calls CachePayload::span.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Updated since
m_fileData is now a span.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Use span for
m_fileData.

* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Updated since
m_fileData is now a span.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Use span for
m_fileData.

* Source/WTF/wtf/FileSystem.h: Use span for m_fileData. Return a span
from leakHandle.

* Source/WebCore/platform/ShareableResource.cpp:
(WebCore::ShareableResource::wrapInSharedBuffer): Updated since
DataSegment::Provider now takes a Function that returns a span.
(WebCore::ShareableResource::span const): Fixed bug where the span
returned here was bigger than it should be: not past the end of the
shared memory, but possibly past the end of the resource.

* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer): Updated
since DataSegment::Provider now takes a Function that returns a span.
(WebCore::FragmentedSharedBuffer::tryCreateArrayBuffer const):
Removed unneeded cast to unsigned. Use copyTo instead of a hand-written
loop to copy the data from m_segments into the ArrayBuffer.
(WebCore::DataSegment::iterate const): Use span.
(WebCore::FragmentedSharedBuffer::startsWith const): Ditto.
(WebCore::FragmentedSharedBuffer::copyTo const): Use span, memcpySpan,
and subspan instead of pointers.
(WebCore::FragmentedSharedBuffer::operator== const): Use equalSpans.
(WebCore::SharedBuffer::span const): Added.
(WebCore::SharedBuffer::data const): Deleted.
(WebCore::SharedBuffer::operator[] const): Updated since
DataSegment::Provider now takes a Function that returns a span.
(WebCore::SharedBuffer::decoder const): Use span.
(WebCore::DataSegment::span const): Added.
(WebCore::DataSegment::data const): Deleted.
(WebCore::DataSegment::size const): Deleted.
(WebCore::SharedBufferDataView::createSharedBuffer const): Updated
since DataSegment::Provider now takes a Function that returns a span.

* Source/WebCore/platform/SharedBuffer.h: Changed
DataSegment::Provider to a single Function that returns a span instead
of one that returns a data pointer and another that returns a size.

* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::SharedMemory::createSharedBuffer const): Updated
since DataSegment::Provider now takes a Function that returns a span.

* Source/WebCore/platform/cocoa/SharedBufferCocoa.mm:
(WebCore::FragmentedSharedBuffer::createCMBlockBuffer const):
Removed unneeded cast.

* Source/WebCore/platform/glib/SharedBufferGlib.cpp:
(WebCore::SharedBuffer::createGBytes const): Use span.

* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
(copyFromSharedBuffer): Deleted.
(WebCore::ScalableImageDecoder::create): Use FragmentedSharedBuffer::copyTo
instead of a local copyFromSharedBuffer function here that does the same thing.

* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::writeFilePathsOrDataBuffersToFile): Use the correct int64_t type here,
which matches the return type of FileSystem::writeToFile.

* Source/WebCore/platform/skia/SharedBufferSkia.cpp:
(WebCore::SharedBuffer::createSkData const): Use span.

* Source/WebKit/NetworkProcess/Notifications/Cocoa/WebPushDaemonConnectionCocoa.mm:
(WebKit::WebPushD::Connection::performSendWithAsyncReplyWithoutUsingIPCConnection const):
Tweak coding style.

* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCocoa.mm:
(WebKit::NetworkCache::Data::adoptMap): Updated since MappedFileData::leakHandle now
returns a span.

* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(dataReferenceFrom): Updated since DataSegment::Provider now takes a Function that
returns a span. This new version now holds a reference to the string instead of
making a copy of the bytes, which seems what was intended from the start.

* Source/WebKit/UIProcess/API/Cocoa/WKBrowsingContextController.mm:
(-[WKBrowsingContextController loadAlternateHTMLString:baseURL:forUnreachableURL:]):
Updated for changes to DataSegment. This now makes the object lifetime clearer
using an explicit RetainPtr.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _loadAlternateHTMLString:baseURL:forUnreachableURL:]): Ditto.

* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::connectionEventHandler): Use auto* for style consistency.

Canonical link: <a href="https://commits.webkit.org/284419@main">https://commits.webkit.org/284419@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/505ed264e2ad6393b300f6be3a5dc9882352230f

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/69313 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/48713 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21986 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/73395 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/20471 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/71430 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/56514 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/20321 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/55128 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13590 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/72379 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/44448 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/59844 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/35606 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/41114 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/17274 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18848 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/62429 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/63056 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/17619 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/75107 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/68559 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/13295 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16842 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62795 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/13334 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59927 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62701 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/15386 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10717 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/4328 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/90341 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/44517 "Built successfully") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/16019 "Found 29 new JSC stress test failures: jsc-layout-tests.yaml/js/script-tests/dfg-compare-final-object-to-final-object-or-other-when-both-proven-final-object.js.layout, jsc-layout-tests.yaml/js/script-tests/dfg-compare-final-object-to-final-object-or-other-when-proven-final-object.js.layout, stress/block-scope-redeclarations.js.dfg-eager, stress/do-eval-virtual-call-correctly.js.bytecode-cache, stress/get-private-name-with-primitive.js.dfg-eager, stress/hashbang.js.bytecode-cache, stress/intl-locale-exceptions.js.dfg-eager, stress/iterator-prototype-find.js.dfg-eager, stress/iterator-prototype-reduce.js.dfg-eager, stress/private-accesor-duplicate-name-early-errors.js.bytecode-cache ... (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/45591 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/46786 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/45332 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/JavaScriptCore/runtime/CachedTypes.cpp:
```diff
@@ -303,9 +303,9 @@ size_t Decoder::size() const
 ptrdiff_t Decoder::offsetOf(const void* ptr)
 {
-    const uint8_t* addr = static_cast<const uint8_t*>(ptr);
+    auto addr = static_cast<const uint8_t*>(ptr);
```
Not `auto*`?

darinadler replied:
Sure, `auto*` it is

cdumez commented on Source/JavaScriptCore/runtime/CachedTypes.cpp:
```diff
@@ -2612,26 +2612,26 @@ RefPtr<CachedBytecode> encodeFunctionCodeBlock(VM& vm, const UnlinkedFunctionCod
 {
-    const auto* cachedEntry = bitwise_cast<const GenericCacheEntry*>(cachedBytecode->span().data());
-    Ref<Decoder> decoder = Decoder::create(vm, WTFMove(cachedBytecode), &key.source().provider());
+    auto cachedEntry = bitwise_cast<const GenericCacheEntry*>(cachedBytecode->span().data());
```
ditto.

cdumez commented on Source/JavaScriptCore/runtime/CachedTypes.cpp:
```diff
@@ -2612,26 +2612,26 @@ RefPtr<CachedBytecode> encodeFunctionCodeBlock(VM& vm, const UnlinkedFunctionCod
         return false;
-    const auto* cachedEntry = bitwise_cast<const GenericCacheEntry*>(cachedBytecode->span().data());
-    Ref<Decoder> decoder = Decoder::create(vm, WTFMove(cachedBytecode));
+    auto cachedEntry = bitwise_cast<const GenericCacheEntry*>(span.data());
```
ditto.

cdumez commented on Source/WTF/wtf/FileSystem.cpp:
```diff
@@ -362,14 +359,11 @@ bool MappedFileData::mapFileHandle(PlatformFileHandle handle, FileOpenMode openM
-    void* data = mmap(0, size, pageProtection, MAP_FILE | (mapMode == MappedFileMode::Shared ? MAP_SHARED : MAP_PRIVATE), fd, 0);
-
-    if (data == MAP_FAILED) {
+    auto data = mmap(0, size, pageProtection, MAP_FILE | (mapMode == MappedFileMode::Shared ? MAP_SHARED : MAP_PRIVATE), fd, 0);
```
ditto.

cdumez commented on Source/WebCore/platform/SharedBuffer.cpp:
```diff
@@ -483,7 +475,7 @@ bool FragmentedSharedBuffer::operator==(const FragmentedSharedBuffer& other) con
         size_t remaining = std::min(thisRemaining, otherRemaining);
 
-        if (memcmp(thisSegment.data() + thisOffset, otherSegment.data() + otherOffset, remaining))
+        if (memcmp(thisSegment.span().data() + thisOffset, otherSegment.span().data() + otherOffset, remaining))
```
Could you have WTF's equalSpans()

darinadler replied:
Yes, will do.

cdumez commented on Source/WebKit/NetworkProcess/Notifications/Cocoa/WebPushDaemonConnectionCocoa.mm:
```diff
@@ -73,7 +73,7 @@
 
         size_t dataSize { 0 };
-        const uint8_t* data = static_cast<const uint8_t *>(xpc_dictionary_get_data(reply, WebPushD::protocolEncodedMessageKey, &dataSize));
+        auto data = static_cast<const uint8_t*>(xpc_dictionary_get_data(reply, WebPushD::protocolEncodedMessageKey, &dataSize));
```
`auto*`?

cdumez reviewed (approved):
LGTM

darinadler committed:
Use span for MappedFileData and DataSegment
https://bugs.webkit.org/show_bug.cgi?id=274756
rdar://128807880

Reviewed by Chris Dumez.

Changed JSC::CachePayload, WTF::MappedFileData, and
WebCore::DataSegment::Provider to use spans instead of pointer/size pairs.

* Source/JavaScriptCore/runtime/CachePayload.cpp:
(JSC::CachePayload::span const): Added.
(JSC::CachePayload::data const): Deleted.
(JSC::CachePayload::size const): Deleted.
* Source/JavaScriptCore/runtime/CachePayload.h: Updated for the above.

* Source/JavaScriptCore/runtime/CachedTypes.cpp:
(JSC::Decoder::offsetOf): Tweaked coding style a bit
(JSC::decodeCodeBlockImpl): Ditto.
(JSC::isCachedBytecodeStillValid): Removed use of CachePayload::size
in function that already calls CachePayload::span.

* Source/WTF/wtf/FileSystem.cpp:
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Updated since
m_fileData is now a span.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Use span for
m_fileData.

* Source/WTF/wtf/win/FileSystemWin.cpp:
(WTF::FileSystemImpl::MappedFileData::~MappedFileData): Updated since
m_fileData is now a span.
(WTF::FileSystemImpl::MappedFileData::mapFileHandle): Use span for
m_fileData.

* Source/WTF/wtf/FileSystem.h: Use span for m_fileData. Return a span
from leakHandle.

* Source/WebCore/platform/ShareableResource.cpp:
(WebCore::ShareableResource::wrapInSharedBuffer): Updated since
DataSegment::Provider now takes a Function that returns a span.
(WebCore::ShareableResource::span const): Fixed bug where the span
returned here was bigger than it should be: not past the end of the
shared memory, but possibly past the end of the resource.

* Source/WebCore/platform/SharedBuffer.cpp:
(WebCore::FragmentedSharedBuffer::FragmentedSharedBuffer): Updated
since DataSegment::Provider now takes a Function that returns a span.
(WebCore::FragmentedSharedBuffer::tryCreateArrayBuffer const):
Removed unneeded cast to unsigned. Use copyTo instead of a hand-written
loop to copy the data from m_segments into the ArrayBuffer.
(WebCore::DataSegment::iterate const): Use span.
(WebCore::FragmentedSharedBuffer::startsWith const): Ditto.
(WebCore::FragmentedSharedBuffer::copyTo const): Use span, memcpySpan,
and subspan instead of pointers.
(WebCore::FragmentedSharedBuffer::operator== const): Use equalSpans.
(WebCore::SharedBuffer::span const): Added.
(WebCore::SharedBuffer::data const): Deleted.
(WebCore::SharedBuffer::operator[] const): Updated since
DataSegment::Provider now takes a Function that returns a span.
(WebCore::SharedBuffer::decoder const): Use span.
(WebCore::DataSegment::span const): Added.
(WebCore::DataSegment::data const): Deleted.
(WebCore::DataSegment::size const): Deleted.
(WebCore::SharedBufferDataView::createSharedBuffer const): Updated
since DataSegment::Provider now takes a Function that returns a span.

* Source/WebCore/platform/SharedBuffer.h: Changed
DataSegment::Provider to a single Function that returns a span instead
of one that returns a data pointer and another that returns a size.

* Source/WebCore/platform/SharedMemory.cpp:
(WebCore::SharedMemory::createSharedBuffer const): Updated
since DataSegment::Provider now takes a Function that returns a span.

* Source/WebCore/platform/cocoa/SharedBufferCocoa.mm:
(WebCore::FragmentedSharedBuffer::createCMBlockBuffer const):
Removed unneeded cast.

* Source/WebCore/platform/glib/SharedBufferGlib.cpp:
(WebCore::SharedBuffer::createGBytes const): Use span.

* Source/WebCore/platform/image-decoders/ScalableImageDecoder.cpp:
(copyFromSharedBuffer): Deleted.
(WebCore::ScalableImageDecoder::create): Use FragmentedSharedBuffer::copyTo
instead of a local copyFromSharedBuffer function here that does the same thing.

* Source/WebCore/platform/network/BlobRegistryImpl.cpp:
(WebCore::writeFilePathsOrDataBuffersToFile): Use the correct int64_t type here,
which matches the return type of FileSystem::writeToFile.

* Source/WebCore/platform/skia/SharedBufferSkia.cpp:
(WebCore::SharedBuffer::createSkData const): Use span.

* Source/WebKit/NetworkProcess/Notifications/Cocoa/WebPushDaemonConnectionCocoa.mm:
(WebKit::WebPushD::Connection::performSendWithAsyncReplyWithoutUsingIPCConnection const):
Tweak coding style.

* Source/WebKit/NetworkProcess/cache/NetworkCacheDataCocoa.mm:
(WebKit::NetworkCache::Data::adoptMap): Updated since MappedFileData::leakHandle now
returns a span.

* Source/WebKit/UIProcess/API/C/WKPage.cpp:
(dataReferenceFrom): Updated since DataSegment::Provider now takes a Function that
returns a span. This new version now holds a reference to the string instead of
making a copy of the bytes, which seems what was intended from the start.

* Source/WebKit/UIProcess/API/Cocoa/WKBrowsingContextController.mm:
(-[WKBrowsingContextController loadAlternateHTMLString:baseURL:forUnreachableURL:]):
Updated for changes to DataSegment. This now makes the object lifetime clearer
using an explicit RetainPtr.
* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _loadAlternateHTMLString:baseURL:forUnreachableURL:]): Ditto.

* Source/WebKit/webpushd/WebPushDaemon.mm:
(WebPushD::WebPushDaemon::connectionEventHandler): Use auto* for style consistency.

Canonical link: https://commits.webkit.org/284419@main
